### PR TITLE
Global and per-source gpgcheck flags

### PR DESCRIPTION
**Pull request by ydirson**
Those two `gpgcheck` and `repo-gpgcheck` flags, both declined in boot-parameter variant (`--no-gpgcheck`) and answerfile variant (`<installation gpgcheck="false">`), are meant as a replacement for the single `<installation netinstall-gpg-check>` parameter in xcp-ng 8.2.1, which was part of c86d2ac1db8a15d537fb9cea9e629a868d318ecf.

- more accurate naming than the original (was never restricted to netinstall)
- more fine-grained control, no need to remove checking of rpms when you just add new or update signed RPMs into the repo

### CODE REVIEW

**Comment by stormi** [comment]
> more accurate naming than the original (was never restricted to netinstall)

I don't think this is true. In 8.2, we only checked signatures in the context of netinstall, AFAIK.

File: answerfile.py [hunk 1]
```diff
@@ -101,6 +101,8 @@ def processAnswerfile(self):
             else:
                 raise AnswerfileException("Unknown mode, %s" % install_type)
 
+            results['gpgcheck'] = getBoolAttribute(self.top_node, ['gpgcheck'], default = True)
+            results['repo-gpgcheck'] = getBoolAttribute(self.top_node, ['repo-gpgcheck'], default = True)
```
Comment (Wescoeur) [inline answerfile.py]: Regarding the coding style, spaces are not used to reference a named param. Here `default = True`. ;)

File: backend.py [hunk 2]
```diff
@@ -426,6 +426,13 @@ def add_repos(main_repositories, update_repositories, repos):
     if not main_repositories or main_repositories[0].identifier() != MAIN_REPOSITORY_NAME:
         raise RuntimeError("No main repository found")
 
+    if 'gpgcheck' in answers and not answers['gpgcheck']:
```
Comment (Wescoeur) [inline backend.py]: Why not directly `not answers.get('gpgcheck')`?

File: backend.py [hunk 3]
```diff
@@ -426,6 +426,13 @@ def add_repos(main_repositories, update_repositories, repos):
...
+        for repo in main_repositories: # update_repositories currently do not check anyway
+            repo.disableGpgCheck()
+    if 'repo-gpgcheck' in answers and not answers['repo-gpgcheck']:
```
Comment (Wescoeur) [inline backend.py]: Same here.

File: doc/answerfile.txt [hunk 4]
```diff
@@ -33,6 +33,28 @@ Restore:
...
 </restore>
 
+
+Common Attributes
+-----------------
+
+  repo-gpgcheck="false"
+
+    Disable check of repodata signature, for all yum repositories
+    (`repo_gpgcheck=0` in `yum.conf`).  Don't use this for a network
```
Comment (Wescoeur) [inline doc/answerfile.txt]: Double spaces. ;)

File: install.py [hunk 5]
```diff
@@ -145,6 +145,14 @@ def go(ui, args, answerfile_address, answerfile_script):
         elif opt == "--netinstall":
             results['netinstall'] = True
             logger.log("This is a netinstall.")
+        # XCP-ng: no-repo-gpgcheck
+        elif opt == "--no-repo-gpgcheck":
+            results['repo-gpgcheck'] = False
+            logger.log("Commandline: no gpg check for yum repository.")
```
Comment (Wescoeur) [inline install.py]: Well, `Commandline:` prefix is not used in the other cases. It doesn't seem very symmetrical. :slightly_smiling_face:

File: repository.py [hunk 6]
```diff
@@ -254,6 +254,8 @@ def __init__(self, accessor):
         super(MainYumRepository, self).__init__(accessor)
         self._identifier = MAIN_REPOSITORY_NAME
         self.keyfiles = []
+        self.__gpg_check = True
+        self.__repo_gpg_check = True
```
Comment (Wescoeur) [inline repository.py]: Seems to me the convention in this project is to use a single "_" for private attributes.

File: repository.py [hunk 7]
```diff
@@ -324,10 +326,12 @@ def _repo_config(self):
...
                 outfh.write(infh.read())
                 return """
-gpgcheck=1
-repo_gpgcheck=1
+gpgcheck=%s
+repo_gpgcheck=%s
 gpgkey=file://%s
-""" % (key_path)
+""" % (1 if self.__gpg_check else 0,
+       1 if self.__repo_gpg_check else 0,
```
Comment (Wescoeur) [inline repository.py]: Can be simplified using `int(self.__gpg_check)`.

**Comment by ydirson** [comment]
> > more accurate naming than the original (was never restricted to netinstall)
> 
> I don't think this is true. In 8.2, we only checked signatures in the context of netinstall, AFAIK.

I could see no explicit restriction in the original code, AFAICT I see no reason why it would not impact installs from local media.

File: answerfile.py (hunk 1 above)
Comment (ydirson) [reply answerfile.py]: Right, I should not have kept that space from the 8.2.1 code  ;)

File: backend.py (hunk 2 above)
Comment (ydirson) [reply backend.py]: Just hesitation on my side - you're right it's just better

File: doc/answerfile.txt (hunk 4 above)
Comment (ydirson) [reply doc/answerfile.txt]: Double spaces after full stop are a reflex here - OK, not in this file :disappointed:

File: install.py (hunk 5 above)
Comment (ydirson) [reply install.py]: Right, I felt that the other messages were not quite informative enough.  In fact the only upstream log on a boot parameter is `Keymap specified on command-line`, so for consistency that should rather be something like `Yum repository gpg check disabled on command-line`  (and we should improve the messages in our other patches).

File: repository.py (hunk 6 above)
Comment (ydirson) [reply repository.py]: I agree we should use a single underscore here, but because they are not "attributes that you do not want subclasses to use" (PEP8 wording)

File: repository.py (hunk 7 above)
Comment (ydirson) [reply repository.py]: right

File: backend.py (hunk 2 above)
Comment (ydirson) [reply backend.py]: In fact there was a reason: `get()` returns `None` when the parameter was never specified, which would be taken as `False`, whereas the default we want is `True`.  Will add a comment to make this more explicit.

Comment (stormi) [reply backend.py]: I suppose we could use `get(..., True)`.

Comment (ydirson) [reply backend.py]: right, that's better

File: install.py [hunk 8]
```diff
@@ -145,6 +145,14 @@ def go(ui, args, answerfile_address, answerfile_script):
         elif opt == "--netinstall":
             results['netinstall'] = True
             logger.log("This is a netinstall.")
+        # XCP-ng: no-repo-gpgcheck
```
Comment (stormi) [inline install.py]: I would remove this comment and the next one, especially if we plan to contribute this upstream (without the comments) and then merge upstream back to our fork.

File: install.py [hunk 9]
```diff
@@ -145,6 +145,14 @@ def go(ui, args, answerfile_address, answerfile_script):
...
+        elif opt == "--no-repo-gpgcheck":
```
Comment (stormi) [inline install.py:139]: The boot parameters (and possibly the options?) should be documented in `doc/parameters.txt`

Comment (ydirson) [reply install.py:139]: Added in new revision

Comment (stormi) [reply install.py:139]: I mentioned the boot parameters in priority, rather that the options handled by init.py, as that's the most useful for users. But I'm not sure what the logic is to decide what options gets described in the boot parameters and what options are described in the script parameters.

Comment (ydirson) [reply install.py:139]: It looks like there is all installer-specific option are described under "Installer", though the "Startup" section is a mix of options interpreted by kernel and scripts launched directly by `init` (I mean `systemd` here, not the installer's own `init` script :dizzy_face:)

**Review by stormi — APPROVED** [review]

**Comment by stormi** [comment]
Are the last pushes since my approval just a rebase, or are there other changes?

**Comment by ydirson** [comment]
> Are the last pushes since my approval just a rebase, or are there other changes?

There is a bit more, especially rework of the `backend.py` changes to prepare better for upcoming patches

File: backend.py [hunk 10]
```diff
@@ -402,6 +402,13 @@ def add_repos(main_repositories, update_repositories, repos):
     if 'sources' in answers_pristine:
         for i in answers_pristine['sources']:
             repos = repository.repositoriesFromDefinition(i['media'], i['address'])
+            repo_gpgcheck = (answers.get('repo-gpgcheck', True) if i['repo_gpgcheck'] is None
+                             else i['repo_gpgcheck'])
+            gpgcheck = (answers.get('gpgcheck', True) if i['gpgcheck'] is None
+                        else i['gpgcheck'])
+            for repo in repos:
+                repo.setRepoGpgCheck(repo_gpgcheck)
+                repo.setGpgCheck(gpgcheck)
```
Comment (stormi) [inline backend.py]: This looks like a partial stepback from what was done in the first two commits. Would it make sense to change the first two commits direclty so that the third one doesn't have to move the code around?

Comment (ydirson) [reply backend.py]: Yes it does.

Also, moving the code from a loop on `main_repositories` into a loop over all sources, as currently done, has the side effect of applying the `*gpgcheck` parameters to all sources, which is likely to cause issues when we suddenly perform the check on an update repo (since XenServer does not activate this check on those).

**Review by benjamreis — APPROVED** [review]

File: backend.py [hunk 11]
```diff
@@ -403,8 +403,10 @@ def add_repos(main_repositories, update_repositories, repos):
...
             repos = repository.repositoriesFromDefinition(i['media'], i['address'])
             add_repos(main_repositories, update_repositories, repos)
-            repo_gpgcheck = answers.get('repo-gpgcheck', True)
-            gpgcheck = answers.get('gpgcheck', True)
+            repo_gpgcheck = (answers.get('repo-gpgcheck', True) if i['repo_gpgcheck'] is None
+                             else i['repo_gpgcheck'])
+            gpgcheck = (answers.get('gpgcheck', True) if i['gpgcheck'] is None
+                        else i['gpgcheck'])
             for repo in repos:
                 if repo in main_repositories:
```
Comment (stormi) [inline backend.py:411]: Are all sources mentioned in the answerfile as a `<source>` in `main_repositories`?

File: answerfile.py [hunk 12]
```diff
@@ -267,7 +269,21 @@ def parseSource(self):
             if rtype == 'url':
                 address = util.URL(address)
 
-            results['sources'].append({'media': rtype, 'address': address})
+            # workaround getBoolAttribute() not allowing "None" as
+            # default, by using a getStrAttribute() call first to
+            # handle the default situation where the attribute is not
+            # specified
+            repo_gpgcheck = (None if getStrAttribute(i, ['repo-gpgcheck'], default=None) is None
+                             else getBoolAttribute(i, ['repo-gpgcheck']))
```
Comment (Wescoeur) [inline answerfile.py:277]: Why not:
```python
getStrAttribute(i, ['repo-gpgcheck'], default=None) and getBoolAttribute(i, ['repo-gpgcheck'])
```
?

Or:
```python
getBoolAttribute(getStrAttribute(i, ['repo-gpgcheck'], default='False'))
```
From my POV, it seems easier to read. I don't care about performance here for the last proposal. :p

Comment (stormi) [reply answerfile.py:277]: Wait a minute. Doesn't getBoolAttribute already handle None as a special default value?

```python
def getBoolAttribute(el, attrs, default = None):
    mandatory = (default == None)
    val = getStrAttribute(el, attrs, '', mandatory).lower()
    if val == '':
        return default
    return val in ['yes', 'true']
```

Comment (stormi) [reply answerfile.py:277]: Oh, there's this `mandatory` parameter though :thinking:

Comment (stormi) [reply answerfile.py:277]: The workaround is not pretty, but I don't find either of Ronan's proposals more readable, personnally (does the second one actually work?).

File: backend.py (hunk 11 above)
Comment (ydirson) [reply backend.py:411]: Assumption was indeed that `main_repositories` comes from `<source>`'s and `update_repositories` from `<driver-source>`'s.  But that's wrong, as supplemental packs can be specified as `<source>` (as documented in `answerfile.txt`).

In more details:

## The workflow goes:
1. parsing
  * `Answerfile.parseSource()`: `<source>` -> `answers['sources']` (named `results` during parsing, `answers` starting from `backend.performInstallation()`)
  * `Answerfile.parseDriverSource()`: `<driver-source>` -> `answers['extra-repos']`
1. (bis) TUI
  * "load driver" screen adds to `answers['extra-repos']` (and does install packages inside initramfs in a non-obvious process, but that's another story, see `driver.doInteractiveLoadDriver()`)
  * the "supplemental packs installation" step at the end of install process XXXX
3. `Repository` objects creation
  * everyone goes through `repository.repositoriesFromDefinition()` and `add_repos()` regardless of "sources" vs. "extra-repos"
4. `add_repos()` separates `UpdateYumRepository`'s (i.e. those with `update.xml`) from the rest, into resp.  `update_repositories`  and `main_repositories`

Essential info: supplemental packs (which include driver packs) all include an `update.xml` (as generated by the DDK's `build-update` script).

So at this point:
- we know the install media `<source>` has no `update.xml`, so it is classified as `main_repository`
- all supplemental packs, whether from `<driver-source>` or from `<source>`, are classified as `update_repository

## What this means for this patch series:
- no supplemental pack is gpgcheck'd as per upstream code (even though some do include a sig! (e.g. `driver-microsemi-aacraid`)
- `<driver-source>` is supposed to point to supplemental packs only, so not having `*gpgcheck` attributes is not a problem until we want to allow checking sigs for them
- a `<source>` element can point declare a source with both "main repos" and supplemental packs; the latter will just ignore the `*gpgcheck` attributes.  This can probably be the case if we build an install media containing supplemental packs alongside the main repo.

## Bottom line:

* we could make the doc a little more clear about supplemental packs not honoring those attributes
* we probably want to add gpgcheck support to supplemental packs too!

File: answerfile.py (hunk 12 above)
Comment (ydirson) [reply answerfile.py:277]: Regardless of the logic operator to use (I guess you meant `or` ?), that would cause an explicit `False` on the source (meant to override the default of `True`) to be indistinguishable from `None` (meant to use the default, especially important here when not the overridden by the global setting).

Comment (ydirson) [reply answerfile.py:277]: > Oh, there's this `mandatory` parameter though thinking

That's the problem, omitting the `default` parameter means making the attribute mandatory, and this is precisely detected using `None`.

But then, it seems pretty sane that `getBoolAttribute()` can only return a bool and never `None`, so I don't think we can do any better, short of introducing an explicit tristate type, which would clearly look like overkill.

File: backend.py (hunk 11 above)
Comment (stormi) [reply backend.py:411]: Thanks for the detailed explanation.

**Review by stormi — APPROVED** [review]

File: backend.py (hunk 11 above)
Comment (ydirson) [reply backend.py:411]: Updated the doc to document the behaviour wrt. Supplemental Packs.  Maybe we'll want to find a wording to make it sound less scary that Supplemental Packs are not gpgcheck'd ?

**Review by Wescoeur — APPROVED** [review]
